recommendation engine (#183)
xmartlabs/blog: _posts/2022-02-08-intro-to-recommendations-engine.md

---
layout: post
title: How recommendation engines boost your online business
date: '2022-02-07T10:00:00.000-03:00'
author: Martin Barreto
tags: [Recommendation Engine, Recommendation Systems, Recommendation engine challenges]
author_id: mtnBarreto
show: true
category: machine-learning
featured_image: /images/taskflow/taskflow-blogpost.jpg
permalink: /blog/intro-to-product-recommendations-engine
---

> Recommendation engines can rapidly increase online business revenue apart from boosting customer loyalty which is crucial for sustainable growth. In this blog post, we’ll go through the main challenges we faced implementing and deploying such solutions in a big fashion brand as well as pros and cons, approaches to implement recommendation engines, and best practices. 

Reviewer: It is always better to have each sentence in a separate line when writing Markdown blogposts so that reviews can be clearer and more concise.

First things first... what exactly is it?, and how can I take advantage of it and put it to work to drive those sales? Keep reading to find out.

## What's a product recommendation engine?

> If you have twitter, Netflix or buy something on Amazon sometimes, you would probably notice how these platforms intensively use recommendation engines throughout every end-user application touchpoint. For example, twitter suggest who you should follow, content that you might like, and top trendings; Netflix recommend movies and series categorized by genre, top watched movies and so on, and Amazon shows purchasable items everywhere. How many time you *bought a gadget and add a (hand-carved wooden) case to the card just because the e-commerce suggest it and of course you need it.*

Reviewer: ```suggestion
If you have Twitter, Netflix or buy something on Amazon sometimes, you would probably notice how these platforms intensively use recommendation engines throughout every end-user application touchpoint. For example, Twitter suggests who you should follow, content that you might like, and top trending topics; Netflix recommend movies and series categorized by genre, top watched movies and so on, and Amazon shows purchasable items everywhere. How many time you *bought a gadget and add a (hand-carved wooden) case to the card just because the e-commerce suggest it and of course you need it.*
```

> Product recommendation engines is a software component that analyze data about platform end-users to infrer what types of products may be interesting to them. Basically it consider user bahaivour, activity, preferences and feedback to generate personalized recommendations. It also take into account products/items chgaracteristics to generate these recoomendations. 

Reviewer: ```suggestion
Product recommendation engines are software components that analyze data about platform end-users to infer what types of products may be interesting to them. Basically it consider user behavior, activity, preferences and feedback to generate personalized recommendations. It also take into account products/items characteristics to generate these recommendations. 
```

Modern Recommendation engines make use of AI and Deep Learning to infer personalized and contextual recommendation apart of being constanly evaluating and trying to improve its model results. You will get into how it measure accurace shortly... 

## Why your platform need a recommendation engine?

Many businesses have quickly understood the value and importance of solid recommendation systems. Our Machine Learning team has seen this reflected in the number of clients looking for one, and no matter your industry, recommendations systems might be a game-changer for your business's online platforms.

First and most importantly reason to adopt a recommendation engine: **increase revenue.** 

- Customer Loyalty: Recommendation systems enables a high level of personalization that makes the user feels the platform provides actual value.
- Platform engagement: Customer easily discover new items to watch, interact, buy and these makes user spend more time on the platform and as a result consume more.

> Recommendation engines drastically boost conversion metrics, take a look at the image below that should its impact in many popular platforms. 

Reviewer: I don't understand "that should its impact in many popular platforms", what do you mean?

![Don't take our word for it, trust the numbers.](/images/recommendation-engines/REC-SYSTEMS.png)
Don't take our word for it, trust the numbers. 

> > Is it worth investing in a recommendation system for my business? How much time or money will it take? Will I be able to see results quickly? Our experts can answer these and many other questions for you. Schedule a free consultation call to learn more opportunities you might be missing.

Reviewer: `Schedule a free consultation call`, should we mention which email they should contact here? Maybe add a link to this phrase?

Author: there is a contact button at the bottom

It's essential to have a recurrent cycle of testing and evaluating with real users, retraining, and deploying the model. Typically, MLOps automatize these tasks and ensure the model keeps updated with recent data and enhances its results.   

Unfortunately, flexibility benefit comes with some downside; it's more complex to design, train and deploy.  

Input data should be carefully analyzed since it has a significant impact on the accuracy; it's essential to have a strong knowledge of the domain to properly design the model, selecting the correct input data to achieve the results.

In the most typical design, there are no hardcoded rules; the deep learning model learns from the data; we just have to provide precise training data. However, sometimes teams mix hardcoded data with this approach to achieve better results. 

[TensorFlow recommenders](https://www.tensorflow.org/recommenders) is a pretty popular ML solution in the industry that is worth viewing.

## Challenges building a Recommendations Engine (and how to address them)

### Choosing the right strategy

> The north star of a recommendation engine is clear; it ultimately must drive revenue. The tricky question is how to successfully prepare the data, design, train, evaluate and deploy the model? 

Reviewer 2: ```suggestion
The north star of a recommendation engine is clear; it ultimately must drive revenue. The tricky question is how to successfully prepare the data, design, train, evaluate, and deploy the model.
```

> A consultancy company like us will help you navigate these challenges, understand the platform needs and create a feasible project plan.

Reviewer 2: ```suggestion
A consultancy company like Xmartlabs will help you navigate these challenges, understand the platform needs and create a feasible project plan.
```

> 1. **Discovery stage:** We get into your company needs and goals, helping your team discover new opportunities leveraging AI-powered solutions. We also agreed on the metrics to measure its performance.  

Reviewer 2: ```suggestion
1. **Discovery stage:** We get into your company's needs and goals, and help your team discover new opportunities by leveraging AI-powered solutions. We also agree on the metrics to measure its performance.  
```

> 2. **Evaluating available data:** As we explained before, gathering data and tracking users’ activity is vital. According to desired outcomes and goals, we reveal data requirements. 

Reviewer 2: ```suggestion
2. **Evaluating available data:** As we explained before, gathering data and tracking users’ activity is vital. Depending on the desired outcomes and goals, we uncover data requirements.
```

    1. Defining what data is required to provide recommendations. 

>     2. Ensuring the data is provided on time and well structured. 

Reviewer 2: ```suggestion
    2. Ensuring the data is well structured and provided on time. 
```

>     3. Ensuring data storage, workflow, and data update rate are suitable according to project goals.

Reviewer 2: ```suggestion
    3. Ensuring data storage, workflow, and update rate are suitable according to project goals.
```

> 3. **UI improvements:** Enumerating screens that will show recommendations to the user. Evaluate UI stuff that needs change and its impact. Create low fidelity and then hi-fidelity design for each screen. 

Reviewer 2: ```suggestion
3. **UI improvements:** Enumerating screens that will show recommendations to the user. Evaluate UI elements that need change and their impact. Create low fidelity and then high-fidelity design for each screen. 
```

> 4. **Agree on a potential solution:** Our specialists design a candidate solution, propose a technology stack and services to build, train, deploy and monitor the recommendation system. 

Reviewer 2: ```suggestion
4. **Agree on a potential solution:** Our specialists design a possible solution, propose a technology stack and services to build, train, deploy and monitor the recommendation system. 
```

5. **Feasible roadmap to go-to-market fast:** Considering everything below, we create a feasible roadmap and assign a dedicated team to work in each solution layer (Product, Data, ML model, UX). 

### People's tastes change over time

> People's tastes evolve over time, so solutions should adapt the RE to provide newer recommendations according to taste changes. Sometimes the product catalog is constantly updated, but these newly released items have no reviews or ranking. But all these items need to be discovered as they are new products or fresh content.  

Reviewer 2: ```suggestion
People's tastes evolve, so solutions should adapt the RE to provide newer recommendations taking those changes into account. Sometimes the product catalog is constantly updated, but these newly released items have no reviews or ranking. However, all these items need to be discovered as they are new products or fresh content.  
```

> To improve the RE performance, its model should be contantly trained, evaluated, and deployed. Likely some data preparation needs to be done too. MLOps, which are essentially DevOps for the ML space, could handle these tasks and automate most, if not the whole, process.

Reviewer 2: ```suggestion
Its model should be constantly trained, evaluated, and deployed to improve the RE performance. It's likely some data preparation needs to be done too. MLOps, which are essentially DevOps for machine learning, could handle these tasks and automate most of, if not all, the process.
```

> According to business domain and goals, the engineering team should decide when and how frequently the model needs to be updated. Having strategically chosen the metric to measure model performance is crucial to improve it. Strategically defining what metrics should be used to monitor and measure recommendations, accuracy is vital to redeploy a retrained model. The engineering team should know when changes improve the system and what's the cause.

Reviewer 2: ```suggestion
According to the business domains and goals, the engineering team should decide when and how frequently the model needs to be updated. Having strategically chosen the metric to measure model performance is crucial to improve it. Strategically defining what metrics should be used to monitor and measure recommendations, accuracy is vital to redeploy a retrained model. The engineering team should know when changes improve the system and what's the cause.
```

### The product catalog is huge

Certain platforms have a long list of items. eBay and Amazon, for instance, have hundreds of thousands of items to recommend for a particular end-user.

> This makes it difficult for the model to determine which items to recommend. A possible solution could be categorizing the items in families of products reducing the candidate items to recommend. The model suggests a product family to then choose an item within the family.

Reviewer 2: ```suggestion
This makes it difficult for the model to determine which items to recommend. A possible solution could be categorizing the items in families of products, reducing the possible items to recommend. The model suggests a product family to then choose an item within the family.
```

### First-time users

> Making recommendations for first-time users involves extra complexity. Not having its personal preference and tracked activity yet, makes it natural to show trending items, most purchased items, and so on. As the user activity is gathered, the recommendations become more precise and personalized. 

Reviewer 2: ```suggestion
Making recommendations for first-time users involves extra complexity. Not having their personal preferences or tracked activity yet makes show trending items and most purchased items the natural solution. The recommendations become more precise and personalized as the user activity is gathered. 
```

> The more the user interacts with the platform, the better recommendations the system can provide. It's an incremental and continuous improvement process that makes its recommendations increasingly accurate and personalized.

Reviewer 2: ```suggestion
The more the user interacts with the platform, the better the recommendations the system will provide. It's an incremental and continuous improvement process that makes its recommendations increasingly accurate and personalized.
```

Depending on the business domain, it could be okay to recommend the most popular items for other new users. 

### Not considering business insights

> A fashion e-commerce would need to consider the season, eventually some important data like Christmas where the demand for certain items increases. Customer preferences can eventually shift according to external factors anytime. 

Reviewer 2: ```suggestion
Fashion e-commerce needs to consider the season, or some important dates like Christmas, where the demand for certain items increases. Customer preferences can eventually shift according to external factors anytime. 
```

Deep Learning-based recommendation systems can learn these patterns by themselves, but the engineering team can facilitate these business rules explicitly. 

The engineering team can tweak input or output data to generate better results or filter undesired items accordingly. 

### Not able to measure outcomes

If you can't measure it, you can't improve it. As simple as that!

> Recommendation systems must be measured with specific metrics to evaluate ROI, evaluate its success, and update the recommendations engine from time to time. 

Reviewer 2: ```suggestion
Recommendation systems must be measured with specific metrics to evaluate ROI, its success, and update the engine from time to time. 
```

> Evaluating the business value of a recommender system mainly depends on the company’s business strategy and its business domain. 

Reviewer 2: ```suggestion
Evaluating the business value of a recommendation system depends mainly on the company’s business strategy and its business domain. 
```

> The most common metrics could be categorized in:

Reviewer 2: ```suggestion
The most common metrics can be categorized as:
```

> **Click-throught rates:**

Reviewer 2: ```suggestion
**Click-throught rate:**
```

> Basically, when an item is recommended and viewed.... Was it clicked? How often? Notice it strongly depends on where the recommended items are displayed, if there are more recommended items around, etc.

Reviewer 2: ```suggestion
When an item is recommended and viewed... Was it clicked? How often? This strongly depends on where the recommended items are displayed, if there are more recommended items around, etc.
```

**User behavior and engagement:**

> For a streaming platform, the recommended engine makes the audience spend more time watching series or movies?? Is there any increase in "Save to watch later" list? etc. 

Reviewer 2: ```suggestion
For a streaming platform, these metrics should show if the recommendation engine makes the audience spend more time watching series or movies or if there's an increase in the "Save to watch later" list? etc. 
```

**Revenue Indicators:**

Recommendations make users spend more money? Are users adding more items to the cart? Revenue indicators are great to measure overall RE performance. 

**Adoption and conversion Rates:**

> Click-through rates are okay, but we are also interested if these suggested items were purchased, added to the cart, or watched. This indicator reflects whether the recommendation engine is useful to the end-user or not. Otherwise, they can be interested in navigating to the recommendation item but never consume it. It's actually a waste of time for the user. 

Reviewer 2: ```suggestion
Click-through rates are okay, but we are also interested to know if these suggested items were purchased, added to the cart, or watched. This indicator reflects whether the recommendation engine is useful to the end-user or not. Otherwise, they can be interested in navigating to the recommendation item but never consume it, which is a waste of time for the user.
```

### Not considering every end-user touchpoint

Evaluating every change and benefit the end-user would experience by implementing the RE is crucial.
Taking Amazon as an example, several views show recommendations:

- Top sellers and most popular items on the homepage - if the user is not logged in.
- Several screen sections showing “More items to explore” - based on previous purchases, viewed items, etc.
    
    ![More items to explore](/images/recommendation-engines/amazon-re-1.png)
    
    ![“More items to explore 2](/images/recommendation-engines/amazon-re-2.png)

- “Frequently bought together” section in the item view.
    
    ![Frequently bought together](/images/recommendation-engines/amazon-re-3.png)
    
    

> - When added a kindle item to the card

Reviewer 2: ```suggestion
- When added a Kindle item to the card
```

>     ![Added a kindle item to the card](/images/recommendation-engines/amazon-re-4.png)

Reviewer 2: ```suggestion
    ![Added a Kindle item to the card](/images/recommendation-engines/amazon-re-4.png)
```

> - During checkout steps

Reviewer 2: ```suggestion
- During checkout process
```

>     ![During checkout steps](/images/recommendation-engines/amazon-re-5.png)

Reviewer 2: ```suggestion
    ![During checkout process](/images/recommendation-engines/amazon-re-5.png)
```

    
- Email

## Closing thoughts

> If you read this far, you probably have a better understanding of how the recommendation system can boost your business. You would probably appreciate the complexity of building one that requires a wide range of specialists collaborating together. 

Reviewer 2: ```suggestion
If you read this far, you probably have a better understanding of how recommendation systems can boost your business. And you'd probably appreciate the complexity of building one that requires a wide range of specialists collaborating together. 
```

Each product is unique, and the challenges can diverge accordingly. All the challenges we presented were experienced during years of developing these solutions for our clients. 

> In the upcoming blog posts, we will be heading to details on how to implement them, different tools and approaches to solve each of the steps presented below, as well as code and practical examples.

Reviewer 2: ```suggestion
In the upcoming blog posts, we'll be going into details on how to implement them, different tools and approaches to solve each of the steps presented below, as well as code and practical examples.
```
